urleb43757003_Origraena_mas_platform_tgz-pJ8-ori_MainJ8/src/ori/mas/core/DefaultHeart.java

```diff
@@ -0,0 +1,284 @@
+		_world = w;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 * <p>
+	 * In this implementation, no influence sum is done.
+	 * All submitted influences are applied in the order they have been added.
+	 * A classic reimplementing subclass will only call <code>super.pulse()</code> and
+	 * then add its specific behaviour, while a more complex heart will completely
+	 * reimplement it.
+	 * </p>
+	 */
+	@Override
+	public List<Influence> pulse() {
+		//System.out.println("Heart.pulse()");
+		List<Influence> madeInfluences = new Fifo<Influence>();
+		Influence i;
+		while (!(_influences.isEmpty())) {
+			i = _influences.pop();
+			//System.out.println("\tbefore makeInfluence: "+i);
+			i = this.makeInfluence(i);
+			//System.out.println("\tafter makeInfluence: "+i);
+			if (i != null)
+				madeInfluences.add(i);
+		}
+		//System.out.println("\tEmpty!!!");
+		return madeInfluences;
+	}
+
+	@Override 
+	public void submitInfluence(Influence i) {
+		//System.out.println("Heart.submitInfluence(i)");
+		//System.out.println("\ti = "+i);
+		if (i != null)
+			_influences.add(i);
+	}
+
+	protected void submitPrioritaryInfluence(Influence i) {
+		_influences.addFirst(i);
+	}
+
+	protected LinkedList<Influence> influences() {
+		return _influences;
+	}
+
+	/**
+	 * Applies an influence and returns the modified influence.
+	 * <p>An influence may be modify if the heart desires it. For example, a movement
+	 * influence will be modify to stop on the first colliding object from the source to
+	 * its destination.
+	 * The returned instance will then contain the new destination point (which surely be
+	 * the colliding point).</p>
+	 * <p>An influence application may submit new influences. In the previous example, a
+	 * colliding influence will also be submitted to the heart by itself.
+	 * </p>
+	 * <p>
+	 * The default implementation does nothing and returns <code>null</code>.
+	 * </p>
+	 * @param i The influence to apply.
+	 * @return The influence which has really been applied (may be <code>null</code>).
+	 */
+	protected Influence makeInfluence(Influence i) {
+		if (i instanceof MultipleInfluence) {
+			for (Influence j : (MultipleInfluence)i)
+				submitPrioritaryInfluence(j);
+			return i;
+		}
+		if (i instanceof PhysicalInfluence)
+			return this.makeInfluence((PhysicalInfluence)i);
+		if (i instanceof BirthInfluence)
+			return this.makeInfluence((BirthInfluence)i);
+		return null;
+	}
+
+	protected Influence makeInfluence(PhysicalInfluence i) {
+		if (i.target() == null)
+			return null;
+		if (i instanceof MovementInfluence)
+			return this.makeInfluence((MovementInfluence)i);
+		if (i instanceof CollideInfluence)
+			return this.makeInfluence((CollideInfluence)i);
+		if (i instanceof ChangeBodyPropertyInfluence)
+			return this.makeInfluence((ChangeBodyPropertyInfluence)i);
+		if (i instanceof DeathInfluence)
+			return this.makeInfluence((DeathInfluence)i);
+		return null;
+	}
+
+	protected Influence makeInfluence(MovementInfluence i) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 152-181 into a separate method.